l20n/migrate/context.py

```diff
@@ -0,0 +1,210 @@
+        f.close()
+
+        ast, errors = self.ftl_parser.parse(contents)
+
+        if len(errors):
+            logger = logging.getLogger('migrate')
+            for err in errors:
+                logger.error(u'Syntax error in {}: {}'.format(path, str(err)))
+
+        return ast
+
+    def add_reference(self, path, realpath=None):
+        """Add an FTL AST to this context's reference resources."""
+        fullpath = os.path.join(self.reference_dir, realpath or path)
+        self.reference_resources[path] = self.read_ftl_resource(fullpath)
+
+    def add_localization(self, path):
+        """Add an existing localization resource.
+
+        If it's an FTL resource, add an FTL AST.  Otherwise, it's a legacy
+        resource.  Use a compare-locales parser to create a dict of (key,
+        string value) tuples.
+        """
+        fullpath = os.path.join(self.localization_dir, path)
+        if fullpath.endswith('.ftl'):
+            ast = self.read_ftl_resource(fullpath)
+            self.localization_resources[path] = ast
+        else:
+            parser = getParser(fullpath)
+            parser.readFile(fullpath)
+            # Transform the parsed result which is an iterator into a dict.
+            collection = {ent.get_key(): ent for ent in parser}
+            self.localization_resources[path] = collection
+
+    def add_transforms(self, path, transforms):
+        """Define transforms for path.
+
+        Each transform is an extended FTL node with `Transform` nodes as some
+        values.  Transforms are stored in their lazy AST form until
+        `merge_changeset` is called, at which point they are evaluated to real
+        FTL nodes with migrated translations.
+
+        Each transform is scanned for `SOURCE` nodes which will be used to
+        build the list of dependencies for the transformed message.
+        """
+        def get_sources(acc, cur):
+            if isinstance(cur, SOURCE):
+                acc.add((cur.path, cur.key))
+            return acc
+
+        for node in transforms:
+            # Scan `node` for `SOURCE` nodes and collect the information they
+            # store into a set of dependencies.
+            dependencies = fold_ftl(get_sources, node, set())
+            # Set this source as a dependency for the current transform.
+            self.dependencies[(path, node.id.name)] = dependencies
+
+        path_transforms = self.transforms.setdefault(path, [])
+        path_transforms.extend(transforms)
+
+    def get_source(self, path, key):
+        """Get an entity from the localized source.
+
+        Used by the `SOURCE` transform.
+        """
+        if path.endswith('.ftl'):
+            resource = self.localization_resources[path]
+            return get_entity(resource.entities(), key)
+        else:
+            resource = self.localization_resources[path]
+            entity = resource.get(key, None)
+            if entity is not None:
+                return entity.get_val()
```

Comment:
@flodolo That sounds like a reason to use `.val`, actually.  I think we'll want to serialize the translations without these characters escaped.